lib/widgets/actions.dart

```diff
@@ -239,3 +241,70 @@ abstract final class ZulipAction {
     }
   }
 }
+
+/// Methods that act through platform APIs and show feedback in the UI.
+///
+/// The static methods on this class can be thought of as higher-level wrappers
+/// for some of the platform binding methods in [ZulipBinding].
+/// But they don't belong there, because they also interact with widgets
+/// in order to present success or error feedback to the user through the UI.
+abstract final class PlatformActions {
+  /// Copies [data] to the clipboard and shows a popup on success.
+  ///
+  /// Must have a [Scaffold] ancestor.
+  ///
+  /// On newer Android the popup is defined and shown by the platform. On older
+  /// Android and on iOS, shows a [Snackbar] with [successContent].
+  ///
+  /// In English, the text in [successContent] should be short, should start with
+  /// a capital letter, and should have no ending punctuation: "{noun} copied".
+  static void copyWithPopup({
+    required BuildContext context,
+    required ClipboardData data,
+    required Widget successContent,
+  }) async {
+    await Clipboard.setData(data);
+    final deviceInfo = await ZulipBinding.instance.deviceInfo;
+
+    if (!context.mounted) return;
+
+    final shouldShowSnackbar = switch (deviceInfo) {
+      // Android 13+ shows its own popup on copying to the clipboard,
+      // so we suppress ours, following the advice at:
+      //   https://developer.android.com/develop/ui/views/touch-and-input/copy-paste#duplicate-notifications
+      // TODO(android-sdk-33): Simplify this and dartdoc
+      AndroidDeviceInfo(:var sdkInt) => sdkInt <= 32,
+      _                              => true,
+    };
+    if (shouldShowSnackbar) {
+      ScaffoldMessenger.of(context).showSnackBar(
+        SnackBar(behavior: SnackBarBehavior.floating, content: successContent));
+    }
+  }
+
+  /// Opens a URL with [ZulipBinding.launchUrl], with an error dialog on failure.
+  // TODO widget tests
```

Comment:
Now that we have a dedicated test group, we should be able to remove the TODO, unless there is something else that needs tests here?

Reply:
Indeed!